From: Andreas Beckmann <anbe@debian.org>
Date: Thu, 30 Jan 2025 23:59:05 -0500
Subject: Provide local lib80211.h header (removed from kernel 6.13+)

Bug-Debian: https://bugs.debian.org/1057447

---
 src/include/lib80211.h | 70 +++++++++++++++++++++++++++++++++++
 src/include/linuxver.h |  8 +++++
 2 files changed, 78 insertions(+)
diff --git a/src/include/lib80211.h b/src/include/lib80211.h
new file mode 100644
index 0000000..e91fe6e
--- /dev/null
+++ b/src/include/lib80211.h
@@ -0,0 +1,70 @@
+/*
+ * lib80211.h - interface for sending 802.11 frames
+ *
+ * Copyright (c) 2002, 2003, 2006 Sam Leffler, Erran Consulting
+ *
+ * Portions Copyright (c) 2005, 2006 David Young.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef _NET_LIB80211_H
+#define _NET_LIB80211_H
+
+#define IEEE80211_FC0_TYPE_MASK		0x0c
+#define IEEE80211_FC0_TYPE_DATA		0x08
+#define IEEE80211_FC0_TYPE_CTL		0x04
+#define IEEE80211_FC0_TYPE_MGT		0x00
+
+#define IEEE80211_FC0_SUBTYPE_MASK	0xf0
+#define IEEE80211_FC0_SUBTYPE_DATA	0x00
+
+#define IEEE80211_FC1_DIR_MASK		0x03
+#define IEEE80211_FC1_DIR_NODS		0x00
+#define IEEE80211_FC1_DIR_TODS		0x01
+#define IEEE80211_FC1_DIR_FROMDS	0x02
+#define IEEE80211_FC1_DIR_DSTODS		0x03
+
+#define IEEE80211_ADDR_LEN	6
+
+struct ieee80211_frame {
+	u_int8_t    i_fc[2];
+	u_int8_t    i_dur[2];
+	u_int8_t    i_addr1[IEEE80211_ADDR_LEN];
+	u_int8_t    i_addr2[IEEE80211_ADDR_LEN];
+	u_int8_t    i_addr3[IEEE80211_ADDR_LEN];
+	u_int8_t    i_seq[2];
+} __packed;
+
+struct ieee80211_frame_addr4 {
+	u_int8_t    i_fc[2];
+	u_int8_t    i_dur[2];
+	u_int8_t    i_addr1[IEEE80211_ADDR_LEN];
+	u_int8_t    i_addr2[IEEE80211_ADDR_LEN];
+	u_int8_t    i_addr3[IEEE80211_ADDR_LEN];
+	u_int8_t    i_seq[2];
+	u_int8_t    i_addr4[IEEE80211_ADDR_LEN];
+} __packed;
+
+#endif /* _NET_LIB80211_H */
diff --git a/src/include/linuxver.h b/src/include/linuxver.h
index b05bc32..8a0d8c0 100644
--- a/src/include/linuxver.h
+++ b/src/include/linuxver.h
@@ -148,8 +148,12 @@ typedef irqreturn_t(*FN_ISR) (int irq, void *dev_id, struct pt_regs *ptregs);
 #endif
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 29)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 13, 0)
+#include <lib80211.h>
+#else
 #include <net/lib80211.h>
 #endif
+#endif
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 29)
 #include <linux/ieee80211.h>
 #else
